h2o-core/src/main/java/water/api/DatasetServlet.java

```diff
@@ -20,18 +20,17 @@
   protected void doGet(HttpServletRequest request, HttpServletResponse response) {
     String uri = ServletUtils.getDecodedUri(request);
     try {
-      boolean use_hex = false;
       String f_name = request.getParameter("frame_id");
       String hex_string = request.getParameter("hex_string");
+      String escape_quotes_string = request.getParameter("escape_quotes");
       if (f_name == null) {
-        throw new RuntimeException("Cannot find value for parameter \'frame_id\'");
+        throw new RuntimeException("Cannot find value for parameter 'frame_id'");
       }
-      if (hex_string != null && hex_string.toLowerCase().equals("true")) {
-        use_hex = true;
-      }
-
       Frame dataset = DKV.getGet(f_name);
-      InputStream is = dataset.toCSV(new Frame.CSVStreamParams().setHexString(use_hex));
+      Frame.CSVStreamParams parms = new Frame.CSVStreamParams()
+          .setHexString(Boolean.parseBoolean(hex_string))
+          .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
+      InputStream is = dataset.toCSV(parms);
```

Comment:
Looks like the `is; is not closed anywhere, not even in `copyStream`. 

I offer the following patch that simplifies the following block by using try-with-resources:

```patch
Index: h2o-core/src/main/java/water/api/DatasetServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-core/src/main/java/water/api/DatasetServlet.java	(revision 6bf5e494de07c7199d46a6d1c182ae2669af02ef)
+++ h2o-core/src/main/java/water/api/DatasetServlet.java	(date 1582538708720)
@@ -4,7 +4,6 @@
 import water.fvec.Frame;
 import water.server.ServletUtils;
 import water.util.FileUtils;
-import water.util.Log;
 
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
@@ -28,36 +27,25 @@
       }
       Frame dataset = DKV.getGet(f_name);
       Frame.CSVStreamParams parms = new Frame.CSVStreamParams()
-          .setHexString(Boolean.parseBoolean(hex_string))
-          .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
-      InputStream is = dataset.toCSV(parms);
-      response.setContentType("application/octet-stream");
-      // Clean up the file name
-      int x = f_name.length() - 1;
-      boolean dot = false;
-      for (; x >= 0; x--)
-        if (!Character.isLetterOrDigit(f_name.charAt(x)) && f_name.charAt(x) != '_')
-          if (f_name.charAt(x) == '.' && !dot) dot = true;
-          else break;
-      String suggested_fname = f_name.substring(x + 1).replace(".hex", ".csv");
-      if (!suggested_fname.endsWith(".csv"))
-        suggested_fname = suggested_fname + ".csv";
-      f_name = suggested_fname;
-      response.addHeader("Content-Disposition", "attachment; filename=" + f_name);
-      ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
-      OutputStream os = null;
-      try {
-        os = response.getOutputStream();
+              .setHexString(Boolean.parseBoolean(hex_string))
+              .setEscapeQuotes(Boolean.parseBoolean(escape_quotes_string));
+
+      try (final InputStream is = dataset.toCSV(parms); final OutputStream os = response.getOutputStream()) {
+        response.setContentType("application/octet-stream");
+        // Clean up the file name
+        int x = f_name.length() - 1;
+        boolean dot = false;
+        for (; x >= 0; x--)
+          if (!Character.isLetterOrDigit(f_name.charAt(x)) && f_name.charAt(x) != '_')
+            if (f_name.charAt(x) == '.' && !dot) dot = true;
+            else break;
+        String suggested_fname = f_name.substring(x + 1).replace(".hex", ".csv");
+        if (!suggested_fname.endsWith(".csv"))
+          suggested_fname = suggested_fname + ".csv";
+        f_name = suggested_fname;
+        response.addHeader("Content-Disposition", "attachment; filename=" + f_name);
+        ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
         FileUtils.copyStream(is, os, 2048);
-      } finally {
-        if (os != null) {
-          try {
-            os.close();
-          }
-          catch (Exception e) {
-            Log.err(e);
-          }
-        }
       }
     } catch (Exception e) {
       ServletUtils.sendErrorResponse(response, e, uri);
```

Reply:
And when we're at it, maybe using `IOUtils.copy(is, os);` would serve us better than our own implementation with custom fixed buffer size (2k bytes might be too small in many cases ? ).

Just noticed that. It's a minor thing. We don't need to do it in this PR.

Reply:
sounds good, but I would do it in separate PR

Reply:
Roger that sir ! No problem here.